### ISSUE TITLE
Have a page table of contents option available

### ISSUE THREAD

**Issue by mdinger** [rust-lang-nursery/mdBook#58]
One thing the rust book does a lot is have really long pages with many subsections in it. For example, see the [modules page](http://doc.rust-lang.org/nightly/book/crates-and-modules.html). One fix for this is to avoid having long pages but I'm not sure that is a good long term strategy that you can impose on users.

A better fix is to have an in-page TOC on every page (perhaps enabled with a flag). The best example I've seen of this is [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/padding) where both left and right contents collapse when the page is thin. The right side floats and follows you down the page which I think is really nice. Even collapsed though, the TOC moves to the top where it can be vertically collapsed as well. I think it's a really nice design.

To accommodate that, you probably would have to make adjustments to the content page width and perhaps to font size. mdBook content and the left sidebar is currently on the wide side though it looks good for as it is for now.

**Comment by azerupi** [comment]
I am not really convinced about this. The rust book has a really *"flat"* toc, and it would make more sense, to me, to divide longer parts in their own sub-chapters. I wouldn't find it abnormal to end up with a book with 3-4 levels of sub-chapters. It creates a very good structure and makes it easier to read because chapters are shorter. That is also why I made an issue about having an option to collapse the toc because it could get quite long with a lot of sub-chapters.

I am interested though, why would you be against dividing long chapters into sub-chapters?

**Comment by mdinger** [comment]
I'm not opposed to them dividing it up but they may never divide it up. It depends on if they decide it's worth the effort, if it would make it better, and so on. As it is, if they decided to not move to a deeper structured document then users are stuck without a page TOC because the tooling doesn't support it. Supporting it at least (optionally) makes life better for users even if the developers decide to be lazy.

Also, the rust docs are kinda okay which means they might have issues but they'll come up less frequently because they're workable. People point out issues when they're really bad more than when they're decent.

One issue with dividing it up is they don't have search built-in, but they do have firefox search built in so really long pages are kinda workable. If the docs become search enabled then this becomes less of an issue though.

The problem with deeply nested TOC is that while the short examples are great, it's difficult to discover the precise example you desire. Doing it well basically requires good search-ability and really good cross-linking.

For example, rustbyexample has 20 chapters with probably an average of 5 subchapters per chapter, so an estimate of 100 pages to look through. Probably more. So if you're looking for info on `ref`, you search ref and click on [one](http://rustbyexample.com/flow_control/match/destructuring/destructure_pointers.html) of the [two](http://rustbyexample.com/scope/borrow/ref.html) `ref` pages. If what you're looking for isn't in one of those two pages (supposing the answer you're looking for is there somewhere) you basically have to skim everything from top to bottom. The rust book might do a little better by getting you to one of the 5 pages that cover the topic.

Another thing is some people actually [prefer](https://internals.rust-lang.org/t/rustdoc-mockup-for-including-more-structured-content/2540/2) big pages so they would probably consider a page table of contents as an advantage.

Basically though, it's kind of complicated but you're probably trying to cover some middle ground so that users who write long pages get good docs, users who write short pages get good docs, and users who don't care get good docs.

**Comment by azerupi** [comment]
You make a couple of good points.
A couple of drawbacks that I can think of:

1. It takes screen estate. Not such a big deal on big screens, but how do we handle smaller screens? Personally, I like a minimalist design, especially for reading. Another toc may add more distraction.

2. What about short pages? It would be silly to have a page toc for a 2 heading long chapter. When is a page considered long? Making a difference between short and long pages adds complexity.

3. A page toc would encourage longer pages. Longer pages don't have their sub-sections listed in the toc, therefore it becomes more difficult to find a sub-section just by skimming through the toc.

-----------------------

>  If the docs become search enabled then this becomes less of an issue though.

Search is an important feature that I plan to add

At the moment, the best way to search for a specific term is to open the print version. It's basically the whole book on one page.

What if instead of a second sidebar on the right we append the page headers to the existing sidebar under the current chapter?

**Comment by mdinger** [comment]
Search is awesome and is definitely needed. This definitely doesn't bar that.

1. Try the MDN site and watch how the sidebars behave when you shrink the width. I think it's pretty smart because both sides collapse and the right sidebar moves to a collapsed and centered TOC.
2. I wouldn't be concerned with header count. Just make it automatically hidden based on window width I guess. The user only has to use it if the window is wide enough. If there was flag to disable it, a rustbyexample style site would probably disable it isn't written in long form. Maybe there could also be a page flag which disables the page TOC for that page.
3. Include all headers in the page TOC. If the user doesn't want something added, they won't add a header.

I'm not actually sure how nice this would be in reality but it seems reasonable. I'm just concerned that the Rust book isn't going to be updated to get tighter structure so I'm trying to propose a solution which would benefit them and be a reasonable idea. I think the docs of the rust book are pretty decent. The wrappings and usability of it aren't.

Your proposal recommending putting it in the left sidebar is similar to my original [rustdoc](https://internals.rust-lang.org/t/rustdoc-mockup-for-including-more-structured-content/2540) proposal. I'm actually for either but would proceed with a mockup to see how well it actually is in practice.

**Comment by mdinger** [comment]
@steveklabnik Do you see the rust book ever getting smaller pages or will it retain it's long form style?

**Comment by steveklabnik** [comment]
Unclear.

**Comment by ishitatsuyuki** [comment]
Having a TOC sidebar on the right seems a commonplace approach these days. Here are two design samples from GitBook and Docusaurus:

https://redux.js.org/introduction/learning-resources
https://docusaurus.io/docs/en/installation

**Comment by zicklag** [comment]
I agree with @ishitatsuyuki. If the TOC is in the sidebar it makes sure that it is not in the way, but that you can still get to it whenever you need to.

**Comment by Dylan-DPC** [comment]
This is an old issue and some of the facts and opinions might have changed over time. If you need this, kindly make a new issue about it so that we can reevaluate it. Thanks
